### PR TITLE
Handle task migrations correctly

### LINKED ISSUE #31: Replace use of CommitFailedException in StreamThread
This exception is Java-only and is raised if one of the following error codes occurs:

* `UNKNOWN_MEMBER_ID`
* `ILLEGAL_GENERATION`
* `REBALANCE_IN_PROGRESS`

(See [this](https://github.com/apache/kafka/blob/c7ab3efcbe5d34c28e19a5a6a59962c2abfd2235/clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java#L781-L787)).

So, the code [here](https://github.com/wintoncode/winton-kafka-streams/blob/2d695e6c4dc089c7388853aba8e54c625fd17520/winton_kafka_streams/processor/_stream_thread.py#L187) needs changing to detect one of these error codes (available from `KafkaException`).

The codes that Python will see are [here](https://github.com/edenhill/librdkafka/blob/bbcfddc1b0fc1087901fb48d8f234d086ed82410/src/rdkafka.h#L240).

### PR DESCRIPTION
**Pull request by jhickson**
Fixes #31.

This removes the previous use of `CommitFailedException` (which is Java-only) and instead mimics the Java Kafka Streams by translating certain error codes received when committing a task into a `TaskMigratedError`. This error is logged by `StreamThread` but otherwise seen as a signal to "try again".

### CODE REVIEW

**Comment by jhickson** [comment]
Made code backwards-compatible with versions for confluent-kafka-python that lack `INVALID_PRODUCER_EPOCH`.

File: tests/test_kafka_streams.py [hunk 1]
```diff
@@ -14,20 +16,18 @@ class MyTestProcessor(BaseProcessor):
     pass
 
 
-class TestKafkaStreams(unittest.TestCase):
+def test_Given_StreamAlreadyStarted_When_CallStartAgain_Then_RaiseError():
```
Comment (llawall) [inline tests/test_kafka_streams.py]: Can we stick with more Pythonic names (separated by underscores)?

Comment (jhickson) [reply tests/test_kafka_streams.py]: I corrected others but missed this - will fix.

File: winton_kafka_streams/kafka_streams.py [hunk 2]
```diff
@@ -7,9 +7,9 @@
 import threading
 from enum import Enum
 
-from .processor import StreamThread
+from winton_kafka_streams.errors.kafka_streams_error import KafkaStreamsError
```
Comment (llawall) [inline winton_kafka_streams/kafka_streams.py]: Any reason not to use relative imports like elsewhere (here and a couple of other modules too)?

Comment (jhickson) [reply winton_kafka_streams/kafka_streams.py]: I thought I'd corrected these. Evidently I missed at least one. I think this is auto-generated by PyCharm.

File: tests/processor/test_stream_task.py [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+from winton_kafka_streams.errors.task_migrated_error import TaskMigratedError
+from winton_kafka_streams.processor import TopologyBuilder
+from winton_kafka_streams.processor._stream_task import StreamTask
+from winton_kafka_streams.processor.task_id import TaskId
+
+taskMigratedErrorCodes = [KafkaError.ILLEGAL_GENERATION,
+                          KafkaError.UNKNOWN_MEMBER_ID,
+                          KafkaError.REBALANCE_IN_PROGRESS,
+                          47 # INVALID_PRODUCER_EPOCH - not supported in all versions for Conluent Kafka so just use the explicit code in this test
+                          ]
```
Comment (ah-) [inline tests/processor/test_stream_task.py]: How recent does confluent_kafka need to be?

We're pretty bleeding edge everywhere else, so could be here as well.

File: winton_kafka_streams/errors/task_migrated_error.py [hunk 4]
```diff
@@ -0,0 +1,9 @@
+"""
+Indicates that a task got migrated to another thread.
+Thus, the task raising this exception can be cleaned up and closed as "zombie".
+"""
```
Comment (ah-) [inline winton_kafka_streams/errors/task_migrated_error.py]: Could put this docstring into the class below, then it'll show up nicely in future docs etc.

http://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html

```python
class TaskMigratedError(KafkaStreamsError): 
    """
    Indicates that a task got migrated to another thread.
    Thus, the task raising this exception can be cleaned up and closed as "zombie".
    """
    pass
```

File: winton_kafka_streams/processor/_stream_task.py [hunk 5]
```diff
@@ -59,6 +72,7 @@ def __init__(self, _task_id, _application_id, _partitions, _topology_builder, _c
         self.context = ProcessorContext(self, self.recordCollector, self.topology.state_stores)
 
         self.punctuation_queue = PunctuationQueue(self.punctuate)
+        # TODO: use the configured timestamp extractor.
```
Comment (ah-) [inline winton_kafka_streams/processor/_stream_task.py]: 👍

File: tests/processor/test_stream_task.py (hunk 3 above)
Comment (jhickson) [reply tests/processor/test_stream_task.py]: It seems the version on Travis is older - that's where the tests failed.

File: winton_kafka_streams/errors/task_migrated_error.py (hunk 4 above)
Comment (jhickson) [reply winton_kafka_streams/errors/task_migrated_error.py]: Done.

File: winton_kafka_streams/processor/_stream_task.py (hunk 5 above)
Comment (jhickson) [reply winton_kafka_streams/processor/_stream_task.py]: Maybe I'll look at this next.

**Pull request merged**